### PR TITLE
[feature](decimal)support  cast rounding half up  and div precision increment in decimalv3.

### PR DESCRIPTION
**Pull request by Mryange**
# Proposed changes

## 1
In mainstream databases, casting is typically rounded using rounding half up.
before
```
mysql [(none)]>select cast(1.115 as decimal(16,2));
+-------------------------------+
| CAST(1.115 AS DECIMAL(16, 2)) |
+-------------------------------+
|                          1.11 |
+-------------------------------+ 
```

after
```
mysql [(none)]>select cast(1.115 as decimal(16,2));
+--------------------------------+
| CAST(1.115 AS DECIMAL64(16,2)) |
+--------------------------------+
|                           1.12 |
+--------------------------------+
```
## 2
divPrecisionIncrement is like mysql https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_div_precision_increment  (using in fe.conf)

before
```
mysql [test]>select v1,v2,v1/v2 from dt;
+---------+-------+-------------+
| v1      | v2    | `v1` / `v2` |
+---------+-------+-------------+
| 1.00000 | 3.000 |     0.33333 |
+---------+-------+-------------+
```
after

```
mysql [test]>select v1,v2,v1/v2 from dt ;
+---------+-------+-------------+
| v1      | v2    | `v1` / `v2` |
+---------+-------+-------------+
| 1.00000 | 3.000 | 0.333333333 |
+---------+-------+-------------+
```
The default value for div_precision_increment is 4.

you can change it 

image

```
mysql [test]>select v1,v2,v1/v2 from dt ;
+---------+-------+-------------------+
| v1      | v2    | `v1` / `v2`       |
+---------+-------+-------------------+
| 1.00000 | 3.000 | 0.333333333333333 |
+---------+-------+-------------------+
```

## 3
pick from https://github.com/apache/doris/pull/19926
## Problem summary

Describe your changes.

## Checklist(Required)

* [x] Does it affect the original behavior
* [ ] Has unit tests been added
* [x] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 33.95 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.1 seconds inserted 10000000 Rows, about 126K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523103036_clickbench_pr_148885.html

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.75 seconds
 stream load tsv:          442 seconds loaded 74807831229 Bytes, about 161 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          75.8 seconds inserted 10000000 Rows, about 131K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523153209_clickbench_pr_149141.html

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p1

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.15 seconds
 stream load tsv:          448 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.9 seconds inserted 10000000 Rows, about 125K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230524074316_clickbench_pr_149570.html

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.41 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          26.2 seconds inserted 10000000 Rows, about 381K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230524130515_clickbench_pr_149830.html

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run arm

**Comment by Mryange** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java [hunk 1]
```diff
@@ -62,14 +63,15 @@ public DataType getDataType() throws UnboundException {
     @Override
     public DecimalV3Type getDataTypeForDecimalV3(DecimalV3Type t1, DecimalV3Type t2) {
         int retPercision = t1.getPrecision() + t2.getScale();
-        int retScale = t1.getScale();
+        int retScale = t1.getScale() + t2.getScale()
+                + ConnectContext.get().getSessionVariable().getDivPrecisionIncrement();
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java:67]: We can't get `ConnectContext` for stream load task. Please put this variable in `fe.conf`

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.88 seconds
 stream load tsv:          444 seconds loaded 74807831229 Bytes, about 160 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          25.7 seconds inserted 10000000 Rows, about 389K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230525035658_clickbench_pr_150100.html

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.98 seconds
 stream load tsv:          441 seconds loaded 74807831229 Bytes, about 161 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          25.5 seconds inserted 10000000 Rows, about 392K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230525044959_clickbench_pr_150158.html

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.86 seconds
 stream load tsv:          443 seconds loaded 74807831229 Bytes, about 161 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.0 seconds inserted 10000000 Rows, about 128K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230525235315_clickbench_pr_150825.html

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.37 seconds
 stream load tsv:          446 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          25.5 seconds inserted 10000000 Rows, about 392K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230526135225_clickbench_pr_151370.html

**Comment by Mryange** [comment]
run p1

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.81 seconds
 stream load tsv:          447 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 insert into select:          25.5 seconds inserted 10000000 Rows, about 392K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230526165443_clickbench_pr_151573.html

**Comment by Mryange** [comment]
run p1

**Comment by Mryange** [comment]
run p1

**Comment by Mryange** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java [hunk 2]
```diff
@@ -62,14 +63,15 @@ public DataType getDataType() throws UnboundException {
     @Override
     public DecimalV3Type getDataTypeForDecimalV3(DecimalV3Type t1, DecimalV3Type t2) {
         int retPercision = t1.getPrecision() + t2.getScale();
-        int retScale = t1.getScale();
+        int retScale = t1.getScale() + t2.getScale()
+                + Config.div_precision_increment;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java:67]: should be the first argument scale + div_precision_increment ? 
```suggestion
        int retScale = t1.getScale() + Config.div_precision_increment;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java [hunk 3]
```diff
@@ -62,14 +63,15 @@ public DataType getDataType() throws UnboundException {
     @Override
     public DecimalV3Type getDataTypeForDecimalV3(DecimalV3Type t1, DecimalV3Type t2) {
         int retPercision = t1.getPrecision() + t2.getScale();
-        int retScale = t1.getScale();
+        int retScale = t1.getScale() + t2.getScale()
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java:66]: `retScale` should be `t1.getScale() + Config.div_precision_increment`

**Review by Gabriel39 — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java [hunk 4]
```diff
@@ -62,14 +63,15 @@ public DataType getDataType() throws UnboundException {
     @Override
     public DecimalV3Type getDataTypeForDecimalV3(DecimalV3Type t1, DecimalV3Type t2) {
         int retPercision = t1.getPrecision() + t2.getScale();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java:65]: percision should also update with div_precision_increment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java (hunk 2 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/Divide.java:67]: use a session variable maybe better?

**Review by morrySnow — COMMENTED** [review]
u need to update decimalv3 doc `sql-manual/sql-reference/Data-Types/DECIMALV3`. and legacy planner's code should also be update.

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run p1

**Comment by Mryange** [comment]
run p0

**Comment by Mryange** [comment]
run clickbench

**Comment by Mryange** [comment]
run beut

**Comment by Mryange** [comment]
run feut

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**